include/verilated.cpp

```diff
@@ -3095,3 +3095,19 @@ void VerilatedAssertOneThread::fatal_different() VL_MT_SAFE {
 #endif
 
 //===========================================================================
+// VlDeleter:: Methods
+
+void VlDeleter::deleteAll() {
+    m_assertOne.check();
+    while (true) {
+        {
+            VerilatedLockGuard lock{m_mutex};
+            std::swap(m_newGarbage, m_toDelete);
+            if (m_toDelete.empty()) break;
+        }
+        for (VlClass* const objp : m_toDelete) delete objp;
+        m_toDelete.clear();
```

Comment:
I believe if another thread enters deleteAll before this finishes we'll break.  We could hold m_mutex but that's kind of slow.  I'd suggest have another m_deleteAll mutex that is obtained AFTER m_mutex (so must succeed to prevent priority inversion), and released after the cleanup finishes.

Reply:
P.S. be sure to comment that // m_deleteAll locked after m_mutex to prevent priority inversion

Just in case we reorganize and forget.

Reply:
Doesn't
```cpp
m_assertOne.check();
```
at the start prevent this?

Reply:
Yes!  Note m_assertOne might not be any faster then getting an uncontested lock.

Reply:
Added a second mutex for `m_toDelete`. This way also works if `eval` is called from a different thread than `VlDeleter` constructor.